internal/internal_worker_usage_collector.go

```diff
@@ -0,0 +1,151 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/uber-go/tally"
+	"go.uber.org/cadence/internal/common/metrics"
+	"go.uber.org/zap"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type (
+	workerUsageCollector struct {
+		workerType      string
+		cooldownTime    time.Duration
+		logger          *zap.Logger
+		ctx             context.Context
+		shutdownCh      chan struct{}
+		wg              *sync.WaitGroup // graceful stop
+		cancel          context.CancelFunc
+		metricsScope    tally.Scope
+		emitOncePerHost oncePerHost
+	}
+
+	workerUsageCollectorOptions struct {
+		Enabled      bool
+		Cooldown     time.Duration
+		MetricsScope tally.Scope
+		WorkerType   string
+		EmitOnce     oncePerHost
+	}
+
+	hardwareUsage struct {
+		NumCPUCores     int
+		CPUPercent      float64
+		NumGoRoutines   int
+		TotalMemory     float64
+		MemoryUsedHeap  float64
+		MemoryUsedStack float64
+	}
+
+	oncePerHost interface {
+		Do(func())
+	}
+)
+
+func newWorkerUsageCollector(
+	options workerUsageCollectorOptions,
+	logger *zap.Logger,
+) *workerUsageCollector {
+	if !options.Enabled {
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	return &workerUsageCollector{
+		workerType:      options.WorkerType,
+		cooldownTime:    options.Cooldown,
+		metricsScope:    options.MetricsScope,
+		logger:          logger,
+		ctx:             ctx,
+		cancel:          cancel,
+		wg:              &sync.WaitGroup{},
+		emitOncePerHost: options.EmitOnce,
+		shutdownCh:      make(chan struct{}),
+	}
+}
+
+func (w *workerUsageCollector) Start() {
+	w.wg.Add(1)
+	go func() {
+		defer func() {
+			if p := recover(); p != nil {
+				w.metricsScope.Counter(metrics.WorkerUsageCollectorPanic).Inc(1)
+				topLine := fmt.Sprintf("WorkerUsageCollector panic for workertype: %v", w.workerType)
+				st := getStackTraceRaw(topLine, 7, 0)
+				w.logger.Error("WorkerUsageCollector panic.",
+					zap.String(tagPanicError, fmt.Sprintf("%v", p)),
+					zap.String(tagPanicStack, st))
+			}
+		}()
+		defer w.wg.Done()
```

Comment:
there are a few things problematic about this goroutine closure
1. this `wg.Done()` will be called once goroutine for `go w.runHardwareCollector()` is started. It shouldn't be marked as done until `runHardwareCollector()` terminates so should be moved there
2. no need for a panic recovery here
3. no need for a goroutine to invoke `runHardwareCollector`.